src/test/java/com/microsoft/sqlserver/clientcertauth/ClientCertificateAuthenticationTest.java

```diff
@@ -0,0 +1,172 @@
+                + clientKey + "-encrypted-pkcs8.key;" + "clientKeyPassword=" + clientKeyPassword + ";";
+        try (Connection conn = DriverManager.getConnection(conStr); Statement stmt = conn.createStatement()) {
+            ResultSet rs = stmt.executeQuery("SELECT @@VERSION AS 'SQL Server Version'");
+            rs.next();
+            assertTrue(rs.getString(1).contains(TestResource.getResource("R_microsoft")));
+        }
+    }
+
+    @Test
+    public void pfxTest() throws Exception {
+        String conStr = connectionString + ";clientCertificate=" + clientCertificate + ".pfx;";
+        try (Connection conn = DriverManager.getConnection(conStr); Statement stmt = conn.createStatement()) {
+            ResultSet rs = stmt.executeQuery("SELECT @@VERSION AS 'SQL Server Version'");
+            rs.next();
+            assertTrue(rs.getString(1).contains(TestResource.getResource("R_microsoft")));
+        }
+    }
+
+    @Test
+    public void pfxEncrytedTest() throws Exception {
+        String conStr = connectionString + ";clientCertificate=" + clientCertificate + "-encrypted.pfx;" + "clientKeyPassword="
+                + clientKeyPassword + ";";
+        try (Connection conn = DriverManager.getConnection(conStr); Statement stmt = conn.createStatement()) {
+            ResultSet rs = stmt.executeQuery("SELECT @@VERSION AS 'SQL Server Version'");
+            rs.next();
+            assertTrue(rs.getString(1).contains(TestResource.getResource("R_microsoft")));
+        }
+    }
+    
+    @Test
+    public void pvkTest() throws Exception {
+        String conStr = connectionString + ";clientCertificate=" + clientCertificate + ".cer;" + "clientKey="
+                + clientKey + ".pvk;" + "clientKeyPassword=" + clientKeyPassword + ";";
+        try (Connection conn = DriverManager.getConnection(conStr); Statement stmt = conn.createStatement()) {
+            ResultSet rs = stmt.executeQuery("SELECT @@VERSION AS 'SQL Server Version'");
+            rs.next();
+            assertTrue(rs.getString(1).contains(TestResource.getResource("R_microsoft")));
+        }
+    }
+
+    @Test
+    public void invalidCert() throws Exception {
+        String conStr = connectionString + ";clientCertificate=invalid_path;" + "clientKeyPassword=" + clientKeyPassword
+                + ";";
+        try (Connection conn = DriverManager.getConnection(conStr)) {
+        } catch (SQLServerException e) {
+            assertTrue(e.getMessage().contains(TestResource.getResource("R_invalidPath")));
+        }
+    }
+
+    @Test
+    public void invalidCertPassword() throws Exception {
+        String conStr = connectionString + ";clientCertificate=" + clientCertificate + ".pfx;"
+                + "clientKeyPassword=invalid_password;";
+        try (Connection conn = DriverManager.getConnection(conStr)) {
+        } catch (SQLServerException e) {
+            assertTrue(e.getMessage().contains(TestResource.getResource("R_keystorePassword")));

```

Comment:
Use `R_pvkParseError` from SQLServerResource instead.

Reply:
Why is this resolved?

Reply:
For this error message, the error message isn't coming from the driver, it's coming from Java. So there's no SQLServerResource text to refer here.